src/registry.rs

```diff
@@ -0,0 +1,131 @@
+// Copyright 2014 The Prometheus Authors
+// Copyright 2016 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::sync::{Arc, RwLock};
+use std::io::Write;
+
+use metrics::Collector;
+use errors::{Result, Error};
+
+struct RegistryCore {
+    pub colloctors_by_id: HashMap<u64, Box<Collector>>,
+    pub dim_hashes_by_name: HashMap<String, u64>,
+}
+
+impl RegistryCore {
+    pub fn register(&mut self, c: Box<Collector>) -> Result<()> {
+        // TODO: should simplify later.
+        let id = {
+            let desc = c.desc();
+
+            if let Some(hash) = self.dim_hashes_by_name.get(&desc.fq_name) {
+                if *hash != desc.dim_hash {
+                    return Err(Error::Msg(format!("a previously registered descriptor with the \
+                                                   same fully-qualified name as {:?} has \
+                                                   different label names or a different help \
+                                                   string",
+                                                  desc)));
+                }
+            }
+
+            if self.colloctors_by_id.contains_key(&desc.id) {
+                return Err(Error::AlreadyReg);
+            }
+
+            self.dim_hashes_by_name.insert(desc.fq_name.clone(), desc.dim_hash);
+
+            desc.id
+        };
+
+        self.colloctors_by_id.insert(id, c);
+        Ok(())
+    }
+
+    pub fn unregister(&mut self, c: Box<Collector>) -> Result<()> {
+        let desc = c.desc();
+        if self.colloctors_by_id.remove(&desc.id).is_none() {
+            return Err(Error::Msg(format!("collector {:?} is not registered", desc)));
+        }
+
+        // dim_hashes_by_name is left untouched as those must be consistent
+        // throughout the lifetime of a program.
+        Ok(())
+    }
+}
+
+#[derive(Clone)]
+pub struct Registry {
```

Comment:
All the time. Metrics may be registered at the same time by different threads. Only registered Metrics can be collected.